hybridse/src/vm/runner.cc

```diff
@@ -2985,47 +3035,118 @@ std::shared_ptr<TableHandler> RequestAggUnionRunner::RequestUnionWindow(
+            int num_rows = 0;
+            agg_row_parser->GetValue(row, "num_rows", type::Type::kInt32, &num_rows);
+
+            // FIXME(zhanghao): check cnt and rows_start_preceding meanings
+            int next_incr = num_rows > 0 ? num_rows - 1 : 0;
+            auto range_status = window_range.GetWindowPositionStatus(cnt + next_incr > rows_start_preceding,
+                                                                     ts_start > end, ts_start < start);
+            if ((max_size > 0 && cnt + next_incr >= max_size) || WindowRange::kExceedWindow == range_status) {
+                start_base = ts_end + 1;
+                break;
+            }
+            if (WindowRange::kInWindow == range_status) {
+                update_agg_aggregator(row);
+                cnt += num_rows;
+            }
 
-        const Row& row = agg_it->GetValue();
-        int64_t ts_end = -1;
-        agg_row_parser->GetValue(row, "ts_end", type::Type::kTimestamp, &ts_end);
-        int num_rows = 0;
-        agg_row_parser->GetValue(row, "num_rows", type::Type::kInt32, &num_rows);
-
-        // FIXME(zhanghao): check cnt and rows_start_preceding meanings
-        int next_incr = num_rows > 0 ? num_rows - 1 : 0;
-        auto range_status = window_range.GetWindowPositionStatus(cnt + next_incr > rows_start_preceding, ts_start > end,
-                                                                 ts_start < start);
-        if ((max_size > 0 && cnt + next_incr >= max_size) || WindowRange::kExceedWindow == range_status) {
-            start_base = ts_end + 1;
-            break;
-        }
-        if (WindowRange::kInWindow == range_status) {
-            update_agg_aggregator(row);
-            cnt += num_rows;
-        }
+            start_base = ts_start;
+            agg_it->Next();
+        } else {
+            const int64_t ts_start = agg_it->GetKey();
+
+            // for agg rows has filter_key
+            // max_size check should happen after iterate all agg rows for the same key
+            std::vector<Row> key_agg_rows;
+
+            int total_rows = 0;
+            int64_t ts_end_range = -1;
+            agg_row_parser->GetValue(agg_it->GetValue(), "ts_end", type::Type::kTimestamp, &ts_end_range);
+            while (agg_it->Valid() && ts_start == agg_it->GetKey()) {
+                const Row& drow = agg_it->GetValue();
+
+                std::string filter_val;
+                if (agg_row_parser->IsNull(drow, "filter_key")) {
+                    LOG(ERROR) << "filter_key is null for *_where op";
+                    agg_it->Next();
+                    continue;
+                }
+                if (0 != agg_row_parser->GetString(drow, "filter_key", &filter_val)) {
+                    LOG(ERROR) << "failed to get value of filter_key";
+                    agg_it->Next();
+                    continue;
+                }
+
+                if (prev_ts_start == ts_start && filter_val == prev_filter_val) {
```

Comment:
here we are assuming the rows with same `filter_val` comes together. 

However it may not be true. e.g.,

ts_start | ts_end | filter_val
|-----|-------|------|
0 | 10 | 0
0 | 10 | 1
0 | 10 | 0
0 | 10 | 1

It doesn't work in this case.

Reply:
yes, using `filter_val` as index key should fix this. 

Or we can use a set to workaroud this, just not performant.

Reply:
as a quickfix, maybe first we use a `set` to check the occurrence of `filter_val`

Reply:
fixed, this case should cover https://github.com/4paradigm/OpenMLDB/pull/2255/files#diff-3f6b69529b323ca236ac25ae70c301269579a5ce1442d569066786bf009e6023R2219

Reply:
fixed, this case should cover https://github.com/4paradigm/OpenMLDB/pull/2255/files#diff-3f6b69529b323ca236ac25ae70c301269579a5ce1442d569066786bf009e6023R2219